File: web/src/stores/training.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as api from '../utils/api'

export const useTrainingStore = defineStore('training', () => {
  const isRunning = ref(false)
  const currentEpisode = ref(0)
  const totalEpisodes = ref(0)
  const metrics = ref<Record<string, number>>({})

  async function fetchStatus() {
    try {
      const res = await api.getTrainingStatus()
      isRunning.value = res.data.is_running
      currentEpisode.value = res.data.current_episode
      totalEpisodes.value = res.data.total_episodes
      metrics.value = res.data.metrics || {}
    } catch {
      // ignore
    }
  }

  async function start(numEpisodes = 50000, numPlayers = 4) {
    await api.startTraining({ num_episodes: numEpisodes, num_players: numPlayers })
    isRunning.value = true
    totalEpisodes.value = numEpisodes
  }

  async function stop() {
    await api.stopTraining()
    isRunning.value = false
  }

  return { isRunning, currentEpisode, totalEpisodes, metrics, fetchStatus, start, stop }
})

File: server/routes/training.py
"""训练控制端点。"""

from __future__ import annotations

import asyncio
import threading
from typing import Optional

from fastapi import APIRouter

from server.schemas import TrainingStartRequest, TrainingStatus

router = APIRouter(prefix="/training", tags=["training"])

# 训练状态
_training_thread: Optional[threading.Thread] = None
_training_status = TrainingStatus(is_running=False)


@router.post("/start")
async def start_training(req: TrainingStartRequest):
    """启动训练任务（后台线程）。"""
    global _training_thread, _training_status

    if _training_status.is_running:
        return {"status": "already_running"}

    _training_status = TrainingStatus(
        is_running=True,
        current_episode=0,
        total_episodes=req.num_episodes,
    )

    def _train():
        global _training_status
        try:
            from ai.train import train, parse_args
            import sys

            # 构造命令行参数
            old_argv = sys.argv
            sys.argv = [
                "train",
                "--episodes", str(req.num_episodes),
                "--num-players", str(req.num_players),
                "--lr", str(req.learning_rate),
                "--device", req.device,
            ]
            train(parse_args())
            sys.argv = old_argv
        except Exception as e:
            print(f"Training error: {e}")
        finally:
            _training_status.is_running = False

    _training_thread = threading.Thread(target=_train, daemon=True)
    _training_thread.start()

    return {"status": "started"}


@router.post("/stop")
async def stop_training():
    """停止训练。"""
    global _training_status
    _training_status.is_running = False
    return {"status": "stopping"}


@router.get("/status", response_model=TrainingStatus)
async def get_training_status():
    """获取训练状态。"""
    return _training_status
